src/sources/util/http/query.rs

```diff
@@ -6,26 +6,63 @@ use vector_lib::{
     event::Event,
 };
 
+use crate::sources::http_server::{build_param_matcher, remove_duplicates, HttpConfigParamKind};
+
 pub fn add_query_parameters(
     events: &mut [Event],
     query_parameters_config: &[String],
     query_parameters: &HashMap<String, String>,
     log_namespace: LogNamespace,
     source_name: &'static str,
 ) {
-    for query_parameter_name in query_parameters_config {
-        let value = query_parameters.get(query_parameter_name);
-        for event in events.iter_mut() {
-            if let Event::Log(log) = event {
-                log_namespace.insert_source_metadata(
-                    source_name,
-                    log,
-                    Some(LegacyKey::Overwrite(path!(query_parameter_name))),
-                    path!("query_parameters", query_parameter_name),
-                    crate::event::Value::from(value.map(String::to_owned)),
-                );
+    let query_parameters_config = build_param_matcher(&remove_duplicates(
+        query_parameters_config.to_vec(),
+        "query_parameters",
+    ));
+    for qp in &query_parameters_config.unwrap() {
```

Comment:
Why is this `unwrap` safe?

Reply:
It's a mistake on my part. I misunderstood the nuances between unwrap and '?'.
As far as I can tell, some patterns **would** result in errors.

I've moved the glob parsing to fn `build` (same as `headers`) so they'd fail in a safe manner.

Reply:
Thanks! That makes sense to "compile" the patterns at start up to propagate any errors.